modules/fundamental/src/ai/service/mod.rs

```diff
@@ -205,156 +259,164 @@ impl AiService {
         let executor = AgentExecutor::from_agent(agent).with_memory(memory.clone());
-
-        let answer = executor
+        _ = executor
             .invoke(prompt_args! {
                 "input" => last_message.content.clone(),
             })
             .await
             .map_err(Error::ChainError)?;
-        history.push(Message::new_ai_message(answer.clone()));
-
-        let mut response = ChatState {
-            messages: Vec::new(),
-        };
 
         let memory = memory.lock().await;
-        for mut message in memory.messages() {
-            if message.message_type == langchain_rust::schemas::MessageType::ToolMessage {
-                // hide tool results from the user
-                message.content = "".to_string();
-            }
-            let internal_state = match serde_json::to_vec(&message) {
-                Ok(serialized) => {
-                    // todo: implement data encryption to avoid client side tampering
-                    STANDARD.encode(serialized.as_slice())
-                }
-                Err(e) => return Err(Error::Internal(e.to_string())),
-            };
-            let ch = ChatMessage {
-                content: message.content.clone(),
-                message_type: match message.message_type.clone() {
-                    langchain_rust::schemas::MessageType::HumanMessage => MessageType::Human,
-                    langchain_rust::schemas::MessageType::AIMessage => MessageType::Ai,
-                    langchain_rust::schemas::MessageType::SystemMessage => MessageType::System,
-                    langchain_rust::schemas::MessageType::ToolMessage => MessageType::Tool,
-                },
-                internal_state: Some(internal_state),
-            };
-            response.messages.push(ch);
-        }
+        let history = memory.messages();
 
-        Ok(response)
-    }
+        // add timestamps for all the new messages added by the LLM to the memory.
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        for _i in 0..(history.len() - timestamps.len()) {
+            timestamps.push(now);
+        }
 
-    pub async fn create_conversation<C: ConnectionTrait>(
-        &self,
-        user_id: String,
-        state: Value,
-        summary: String,
-        connection: &C,
-    ) -> Result<conversation::Model, Error> {
-        let model = conversation::ActiveModel {
-            id: Default::default(),
-            user_id: Set(user_id),
-            state: Set(state),
-            seq: Set(0),
-            summary: Set(summary),
-            updated_at: Set(OffsetDateTime::now_utc()),
+        // convert the memory messages to ChatMessages
+        let internal_state = InternalState {
+            messages: history,
+            timestamps,
         };
-        Ok(model.insert(connection).await?)
+        Ok(internal_state)
     }
 
-    pub async fn update_conversation<C: ConnectionTrait>(
+    pub async fn upsert_conversation<C: ConnectionTrait>(
```

Comment:
In the implementation is says that the update can take a while. How will that affect concurrent calls this this function. Which to my understanding can happen when multiple calls via the API happen.

Reply:
Could do.

Reply:
but given seq is part of the data model response object, seems odd to post it via headers.

Reply:
The pattern of `If-Match` and `ETag` seems to be a standard thing.

Reply:
I think that tend to be a thing when the etag is not part of the document.  Example: hash of a file etc.

Reply:
But I guess it can't hurt to use the etag system. It should help with caching too.

Reply:
Why invent something new, if there's an existing pattern?!